src/Compressors/count_sketch.jl

```diff
@@ -0,0 +1,176 @@
+- `carinality::Cardinality`, the direction the compression matrix is intended to be
+    applied to a target matrix or operator. Values allowed are `Left()` or `Right()`.
+    By default `Left()` is chosen.
+- `compression_dim`, the target compression dimension. Referred to as ``s`` in the
+    mathemtical description. By default this is set to 2.
+- `type::Type{<:Number}`, the type of the elements in the compressor. By default is set
+    to Float64.
+
+## Returns
+- A `CountSketch` object.
+
+## Throws
+- `ArgumentError` if `compression_dim` is non-positive
+"""
+struct CountSketch <: Compressor
+    cardinality::Cardinality
+    compression_dim::Int64
+    type::Type{<:Number}
+    # Check on the compression dimension
+    function CountSketch(cardinality, compression_dim, type)
+        if compression_dim <= 0
+            throw(ArgumentError("Field 'compression_dim' must be positive."))
+        end
+
+        return new(cardinality, compression_dim, type)
+    end
+end
+
+function CountSketch(;
+    cardinality::Cardinality=Left(),
+    compression_dim::Int64=2,
+    type::Type{<:Number}=Float64
+)
+    # Partially construct the count sketch datatype
+    return CountSketch(cardinality, compression_dim, type)
+end
+
+"""
+    CountSketchRecipe <: CountSketchRecipe
+
+The recipe containing all allocations and information for the CountSketch compressor.
+
+# Fields
+- `cardinality::C where C<:Cardinality`, the cardinality of the compressor. The
+value is either `Left()` or `Right()`.
+- `compression_dim::Int64`, the target compression dimension.
+- `n_rows::Int64`, the number of rows of the compression matrix.
+- `n_cols::Int64``, the number of columns of the compression matrix.
+- `mat::SparseMatrixCSC`, the compression matrix stored in a sparse form.
+"""
+mutable struct CountSketchRecipe{C<:Cardinality} <: CompressorRecipe
+    cardinality::Cardinality
+    compression_dim::Int64
+    n_rows::Int64
+    n_cols::Int64
+    mat::SparseMatrixCSC
+end
+
+function CountSketchRecipe(
+    cardinality::Left,
+    compression_dim::Int64,
+    A::AbstractMatrix,
+    type::Type{<:Number},
+)
+    # determine the initial size
+    n_rows = compression_dim
+    n_cols = size(A, 1)
+    initial_size = n_cols
+    # assign -1 or +1 in every row/column with probability 0.5
+    signs = convert(Vector{type}, rand([-1.0, 1.0], initial_size))
+    groups = rand(1:compression_dim, initial_size)
+    ptr = collect(1: initial_size)
+    mat = sparse(groups, ptr, signs, n_rows, n_cols)
+    return CountSketchRecipe{Left}(cardinality, compression_dim, n_rows, n_cols, mat)
+end
+
+function CountSketchRecipe(
+    cardinality::Right,
+    compression_dim::Int64,
+    A::AbstractMatrix,
+    type::Type{<:Number},
+)
+    # determine the initial size
+    n_rows = size(A, 2)
+    n_cols = compression_dim
+    initial_size = n_rows
+    # assign -1 or +1 in every row/column with probability 0.5
+    signs = convert(Vector{type}, rand([-1.0, 1.0], initial_size))
```

Comment:
Same comment here. No need to convert the output vector, just the input entries.